File: pcomtest.m
function [P_rn,pval,cval] = pcomtest(p_i,alph,r)

% Function to implement intersection union of 
% Spreng and Urga (2022) Combining p-Values for Multivariate Predictive
% Ability Testing, Journal of Business and Economic Statistics
%
% Input: 
%   - p_i: (n x 1) vector of p-values
%   - alph: significance level (default 5%)
%   - r: parameter for test statistic (default 10)
%
% Output: 
%   - P_rn: test statistic
%   - pval: p-value of test statistic
%   - cvla: critical value of test statistc

if isempty(alph)
    alph = 0.05;
end

if isempty(r)
    r = 10;
end

n = length(p_i);

% Numerical trick in case input pvalues are rounded down to zero (usual practice in most programmes)
% to prevent infinity values
p_i(p_i<=10^(-4)) = 10^(-4); 

P_rn = sum(p_i.^(-r))^(1/r)/n;        % Test statistic
pval = min(r/((r-1)*P_rn),1);           % p-value (proposition 1)
cval = r/(alph*(r-1));                % Critical value

disp('----------------------------------------'), disp('')
disp('Intersection Union Test:'), disp('')
disp('----------------------------------------')

disp(''), disp( sprintf('Test statistic: %1.4f' , P_rn ) ), disp('')
disp(''), disp( sprintf('Critical value: %1.4f' , cval ) ), disp('')
disp(''), disp( sprintf('p-value: %1.4f' , pval ) ), disp('')
disp('----------------------------------------')

if P_rn >= cval
    disp('Conclude: reject H_0')
else
    disp('Conclude: do not reject H_0')
end
disp('----------------------------------------'), disp('')

File: pcomtest.py
"""
Function to implement intersection union of 
Spreng and Urga (2022) Combining p-Values for Multivariate Predictive
Ability Testing, Journal of Business and Economic Statistics
"""
import numpy as np

p_i = np.array([0.05, 0.02, 0.1, 0.0])

def pcom_test(p_i,alpha=0.05,r=10):

    if type(p_i) != np.ndarray:
        raise ValueError("Vector of p-values must be numpy array")

    n = p_i.size
    p_i[p_i<=10**(-4)] = 10**(-4)

    p_rn = np.sum(p_i**(-r))**(1/r)/n      # Test statistic
    pval = np.minimum(r/((r-1)*p_rn),1)    # p-value (proposition 1)
    cval = r/(alpha*(r-1));                # Critical value

    print(
        "---------------------------------------- \n"
        "Intersection Union Test: \n"
        "---------------------------------------- \n"
        "Test Statistic:", p_rn, "\n"
        "Critical Value:", cval, "\n"
        "p-value:", pval, "\n"
        "----------------------------------------"
     )

    if p_rn >= cval:
        print("Conclude: reject H_0")
    else:
        print("Conclude: do not reject H_0")
    print("----------------------------------------")

    return p_rn, pval, cval